artiq/devices/novatech409B/novatech409B.py

```diff
@@ -0,0 +1,362 @@
+                    print("ERROR :: novatech409B.ser_send() "
+                        "response was {}".format(result))
+                    return (result, expected_response)
+                    sys.exit()
+            return (result, expected_response)
+        else:
+            # in simulation mode
+            return (expected_response, expected_response)
+
+    def reset(self):
+        """command hardware reset of 409B
+
+        returns: None
+        """
+        self.debug_message("reset", "", level=3)
+        self.ser_send("R", ignore_unusual_response=True)
+        time.sleep(1)
+        self.setup()
+
+    def setup(self):
+        """initial setup of 409B
+
+        Setup the Novatech 409B with the following defaults.
+        * command echo off ("E d")
+        * external clock ("") 10 MHz sinusoid -1 to +7 dBm
+
+        :returns: None
+        """
+        self.debug_message("setup", "", level=2)
+        #disable command echo
+        self.ser_send("E d", ignore_unusual_response=True)
+        self.set_phase_continuous(True)
+        self.set_simultaneous_update(False)
+
+    def save_state_to_eeprom(self):
+        """save current state to EEPROM
+
+        Saves current state into EEPROM and sets valid flag.
+        State used as default upon next power up or reset. """
+        self.debug_message("save_state_to_eeprom", "", level=2)
+        self.ser_send("S")
+
+    def set_phase_continuous(self, is_continuous):
+        """toggle phase continuous mode
+
+        Sends the “M n” command. This turns off the automatic
+        clearing of the phase register. In this mode, the phase
+        register is left intact when a command is performed.
+        Use this mode if you want frequency changes to remain
+        phase synchronous, with no phase discontinuities.
+
+        :param bool is_continuous: True or False
+        """
+
+        self.debug_message("set_phase_continuous", "", level=2)
+        if is_continuous:
+            self.ser_send("M n")
+        else:
+            self.ser_send("M a")
+
+    def set_simultaneous_update(self, my_bool):
+        """
+        :param bool my_bool: True or False
+
+        Sends the “I m” command. In this mode an update
+        pulse will not be sent to the DDS chip until
+        an “I p” command is sent. This is useful when it is
+        important to change all the outputs to new values
+        simultaneously."""
+        self.debug_message("set_simultaneous_update", "", level=2)
+        if my_bool:
+            self.ser_send("I m")
+        else:
+            self.ser_send("I a")
+
+    def set_freq(self, ch_no, freq):
+        """set_freq(ch_no,freq):
+        Set ch_no to frequency freq MHz"""
+        self.debug_message("set_freq", str(ch_no)+","+str(freq), level=3)
+        if ch_no < 0 or ch_no > 3:
+            print("ERROR :: novatech409B.set_freq() ch_no Error")
+            sys.exit()
+        if freq < 0.0 or freq > 171.1276031:
```

Comment:
The device does not check the range. The number of low-level details which can influence this number are too numerous given my aspiration for this driver. If the device is configured to used it's internal reference servo'd to an external reference via a PLL this is the appropriate constant. I'll let future users worry about other use cases.